mlflow/utils/proto_json_utils.py

```diff
@@ -230,6 +230,13 @@ def cast_df_types_according_to_schema(pdf, schema):
                     pdf[col_name] = pdf[col_name].map(
                         lambda x: base64.decodebytes(bytes(x, "utf8"))
                     )
+                elif schema.is_tensor_spec() and isinstance(pdf[col_name][0], list):
```

Comment:
Is it possible for the first column to be scalar but the second / third columns to be lists?

Reply:
`pdf[col_name][0]` is not checking first column, it checks the first element in `col_name` column

Reply:
Whoops! Sorry :)

Reply:
What if the first element in the column isn't a list but the second / third element is, etc?

Reply:
> What if the first element in the column isn't a list but the second / third element is, etc?

This is rarely happen. I prefer only to check the first element for fast type checking, to avoid too much overhead. If the case "first element in the column isn't a list but the second / third element is" happen, the following reshaping opertion will still raise error (but raise less informative error message)

Reply:
Perfect - as long as it raises an error :)